astropy/stats/funcs.py

```diff
@@ -834,20 +835,24 @@ def median_absolute_deviation(data, axis=None, func=None, ignore_nan=False):
             func = np.nanmedian
         else:
             is_masked = False
-            func = np.median
+            func = np.median  # drops units if result is NaN
     else:
         is_masked = None
 
     data = np.asanyarray(data)
     # np.nanmedian has `keepdims`, which is a good option if we're not allowing
     # user-passed functions here
     data_median = func(data, axis=axis)
+    if isinstance(data, u.Quantity) and func is np.median:
```

Comment:
I'd maybe target the case needed even more, if only to avoid slowing down the more common case where it is not a problem, with:
```
if isinstance(data, u.Quantity) and func is np.median and data.ndim == 0 and np.isnan(data):
```

Also, perhaps add a note to #12165? (I've got a fix in the works...)

Reply:
Yes, that's better.  I also added a comment about #12165.